storage/src/backend/connection.rs

```diff
@@ -428,90 +450,110 @@ impl Connection {
         let current_mirror = self.mirror_state.current.load();
 
         if let Some(mirror) = current_mirror.as_ref() {
-            let data_cloned: Option<ReqBody<R>> = match data.as_ref() {
-                Some(ReqBody::Form(form)) => Some(ReqBody::Form(form.clone())),
-                Some(ReqBody::Buf(buf)) => Some(ReqBody::Buf(buf.clone())),
-                _ => None,
-            };
-
-            let mirror_host = Url::parse(&mirror.config.host).map_err(ConnectionError::Url)?;
-
-            let mut current_url = Url::parse(url).map_err(ConnectionError::Url)?;
-
-            current_url
-                .set_scheme(mirror_host.scheme())
-                .map_err(|_| ConnectionError::Scheme)?;
-            current_url
-                .set_host(mirror_host.host_str())
-                .map_err(|_| ConnectionError::Host)?;
-            current_url
-                .set_port(mirror_host.port())
-                .map_err(|_| ConnectionError::Port)?;
-
-            if let Some(working_headers) = &mirror.config.headers {
-                for (key, value) in working_headers.iter() {
+            // With configuration `auth_through` disabled, we should not intend to send authentication
+            // request to mirror. Mainly because mirrors like P2P/Dragonfly has a poor performance when
+            // relaying non-data requests. But it's still possible that ever returned token is expired.
+            // So mirror might still respond us with status code UNAUTHORIZED, which should be handle
+            // by sending authentication request to the original registry.
+            //
+            // - For non-authentication request with token in request header, handle is as usual requests to registry.
+            //   This request should already take token in header.
+            // - For authentication request
+            //      1. auth_through is disabled(false): directly pass below mirror translations and jump to original registry handler.
+            //      2. auth_through is enabled(true): try to get authenticated from mirror and should also handle status code UNAUTHORIZED.
+            if mirror.config.auth_through
+                || !requesting_auth && headers.contains_key(HEADER_AUTHORIZATION)
```

Comment:
Does the `HEAD /blobs` request also go through mirror server?

Reply:
Yes, in the current control flow, the HEAD /blobs request also goes through the mirror server.
Does the HEAD /blobs request also get a `401` response when the `headers` do not contain the authorization token?

Reply:
> Yes, in the current control flow, the HEAD /blobs request also goes through the mirror server. Does the HEAD /blobs request also get a `401` response when the `headers` do not contain the authorization token?

Yes, this request also needs authorization.